src/skmatter/metrics/_prediction_rigidities.py

```diff
@@ -0,0 +1,199 @@
+    structures are built by concatenating the descriptors of different prediction
+    components together. It also assumes, like the case of LPR, that model training
+    is undertaken in a manner where the global prediction targets are averaged over
+    the number of atoms appearing in each training structure, and the average feature
+    vector of each structure is hence used in the regression. Likewise, to guarantee
+    valid comparison in the (L)CPR across different models, feature vectors are scaled
+    by a global factor based on standard deviation across atomic envs.
+
+    Parameters
+    ----------
+    X_train : list of ndarray of shape (n_atoms, n_features)
+        Training dataset where each training set structure is stored as a separate 
+        ndarray.
+
+    X_test : list of ndarray of shape (n_atoms, n_features)
+        Test dataset where each training set structure is stored as a separate 
+        ndarray.        
+
+    reg : float value
+        Regularizer value that the linear/kernel model has been optimized to.
+
+    comp_dims : array of int values
+        Dimensions of the feature vectors pertaining to each prediction component.
+        
+
+    Returns
+    -------
+    CPR : ndarray of shape (n_test_strucs, n_components)
+    LCPR : list of ndarrays of shape (n_atoms, n_components)
+    rank_diff : integer value of the difference between cov matrix dimension and rank
+        
+    """
+
+    # initialize a StandardFlexibleScaler and fit to train set atom envs
+    scaler = StandardFlexibleScaler(with_mean=False)
+    X_atom = np.vstack(X_train)
+    scaler.fit(X_atom)
+
+    # prep to build covariance matrix XX, take average feat vecs per struc
+    X_struc = []
+    for i, cur_X in enumerate(X_train):
+        X_struc.append(np.mean(scaler.transform(cur_X), axis=0))
+    X_struc = np.vstack(X_struc)
+
+    # build XX and obtain Xinv for LPR calculation
+    XX = X_struc.T @ X_struc
+    Xprime = XX + reg * np.eye(XX.shape[0])
+    rank_diff = X_struc.shape[1] - np.linalg.matrix_rank(Xprime)
+    Xinv = np.linalg.inv(Xprime)
+
+    # track test set atom indices for output
+    lens = []
+    for X in X_test:
+        lens.append(len(X))
+    test_idxs = np.cumsum([0] + lens)    
+
+    # get struc average feat vecs for test set
+    X_struc_test = []
+    for i, cur_X in enumerate(X_test):
+        X_struc_test.append(np.mean(scaler.transform(cur_X), axis=0))
+    X_struc_test = np.vstack(X_struc_test)
+
+    # prep and compute CPR and LCPR
+    num_test = len(X_test)
+    num_comp = len(comp_dims)
+    comp_idxs = np.cumsum([0] + comp_dims.tolist())
+    X_test = np.vstack(X_test)
+    atom_count = X_test.shape[0]
+    CPR = np.zeros((num_test, num_comp))
+    LCPR_np = np.zeros((atom_count, num_comp))
+
+    for ci in range(num_comp):
+
+        tot_comp_idx = np.arange(comp_dims.sum()) 
+        mask = ((tot_comp_idx >= comp_idxs[ci]) & (tot_comp_idx < comp_idxs[ci+1])).astype(float)
```

Comment:
It seems to me that you just take ranges of the matrix in the end. Why don't you just do?
```python
comp_dims = [0] + comp_dims.tolist()
#...
  A = ... # some calcluators
  Xic = A[:, comp_dims[i]:comp_dims[i+1]]
```
I dont know if I am missing something, but at the moment it seems to me a convoluted way to do this.

Reply:
Thanks for the catch. you are right, i was mindlessly coding and had end up with something convoluted.

Reply:
Thanks for the catch. Here however, we must keep the original feature vector length. I will come up with some middle ground..